software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/WinRmExecuteHelper.java

```diff
@@ -0,0 +1,217 @@
+    public static final Logger LOG = LoggerFactory.getLogger(WinRmExecuteHelper.class);
+
+    private Entity entity;
+    private Task<Integer> task;
+
+    protected final NativeWindowsScriptRunner runner;
+    public final String summary;
+
+    private String command;
+    private String psCommand;
+
+    @SuppressWarnings("rawtypes")
+    protected final Map flags = new LinkedHashMap();
+    protected Predicate<? super Integer> resultCodeCheck = Predicates.alwaysTrue();
+    protected ByteArrayOutputStream stdout, stderr, stdin;
+
+    public WinRmExecuteHelper(NativeWindowsScriptRunner runner, String summary) {
+        this.runner = runner;
+        this.summary = summary;
+    }
+
+    public WinRmExecuteHelper setCommand(String command) {
+        this.command = command;
+        return this;
+    }
+
+    public WinRmExecuteHelper setPsCommand(String psCommand) {
+        this.psCommand = psCommand;
+        return this;
+    }
+
+    protected Entity getEntity() {
+        return entity;
+    }
+
+    /** queues the task for execution if we are in a {@link TaskQueueingContext} (e.g. EffectorTaskFactory);
+     * or if we aren't in a queueing context, it will submit the task (assuming there is an {@link ExecutionContext}
+     * _and_ block until completion, throwing on error */
+    @Beta
+    public Task<Integer> queue() {
+        return DynamicTasks.queueIfPossible(newTask()).orSubmitAndBlock().getTask();
+    }
+
+    /** creates a task which will execute this script; note this can only be run once per instance of this class */
+    public synchronized Task<Integer> newTask() {
+        if (task!=null) throw new IllegalStateException("task can only be generated once");
+        TaskBuilder<Integer> tb = Tasks.<Integer>builder().displayName("winrm: "+summary).body(
+                new Callable<Integer>() {
+                    public Integer call() throws Exception {
+                        return executeInternal();
+                    }
+                });
+
+        try {
+            ByteArrayOutputStream stdin = new ByteArrayOutputStream();
+            stdin.write((command != null ? command : psCommand).getBytes());
+            tb.tag(BrooklynTaskTags.tagForStreamSoft(BrooklynTaskTags.STREAM_STDIN, stdin));
+        } catch (IOException e) {
+            LOG.warn("Error registering stream "+BrooklynTaskTags.STREAM_STDIN+" on "+tb+": "+e, e);
+        }
+
+        Map flags = getFlags();
+
+        Map<?,?> env = (Map<?,?>) flags.get("env");
+        if (env!=null) {
+            // if not explicitly set, env will come from getShellEnv in AbstractSoftwareProcessSshDriver.execute,
+            // which will also update this tag appropriately
+            tb.tag(BrooklynTaskTags.tagForEnvStream(BrooklynTaskTags.STREAM_ENV, env));
+        }
+
+        if (gatherOutput) {
+            stdout = new ByteArrayOutputStream();
+            tb.tag(BrooklynTaskTags.tagForStreamSoft(BrooklynTaskTags.STREAM_STDOUT, stdout));
```

Comment:
Why do we now have this code here and in `AbstractSoftwareProcessWinRmDriver.executeCommand`? Is there a way to just have it in one place? e.g. could `AbstractSoftwareProcessWinRmDriver.executeCommand` use the `WinRmExecuteHelper` somehow?